src/views/text_input.rs

```diff
@@ -59,872 +67,545 @@ prop_extractor! {
+            let (content, offset, preedit_range) = {
+                let content = String::from(doc.rope_text().text());
+                if let Some(preedit) = doc.preedit().preedit.get().as_ref() {
+                    let mut new_content = String::new();
+                    new_content.push_str(&content[..offset]);
+                    new_content.push_str(&preedit.text);
+                    new_content.push_str(&content[offset..]);
+                    let range = (offset, offset + preedit.text.len());
+                    let offset = preedit
+                        .cursor
+                        .as_ref()
+                        .map(|(_, end)| offset + *end)
+                        .unwrap_or(offset);
+                    (new_content, offset, Some(range))
+                } else {
+                    (content, offset, None)
+                }
+            };
+            id.update_state(TextInputState::Content {
+                text: content.clone(),
+                offset,
+                preedit_range,
+            });
+        });
+    }
+
+    {
+        create_effect(move |_| {
+            let focus = is_focused.get();
+
+            if focus {
+                // Reset blink interval every time we gain focus
+                editor.with_untracked(|e| e.cursor_info.reset());
+            }
+            id.update_state(TextInputState::Focus(focus));
+        });
+
+        // Syncs with the blinking state of the cursor.
+        // This runs every 500 milliseconds when the input is focused
+        create_effect(move |_| {
+            let focus = is_focused.get();
+            if focus {
+                let hidden = editor.with_untracked(|e| e.cursor_info.hidden);
+                hidden.track();
+                id.request_paint();
+            }
+        });
+
+        let editor = editor.get();
+        let ime_allowed = editor.ime_allowed;
+        create_effect(move |_| {
+            let focus = is_focused.get();
+            if focus {
+                if !ime_allowed.get_untracked() {
+                    ime_allowed.set(true);
+                    set_ime_allowed(true);
+                }
+                let cursor_line = cursor_line.get();
+
+                let window_origin = window_origin.get();
+                let viewport = editor.viewport.get();
+                let origin = window_origin
+                    + Vec2::new(
+                        cursor_line.p1.x - viewport.x0,
+                        cursor_line.p1.y - viewport.y0,
+                    );
+                set_ime_cursor_area(origin, Size::new(800.0, 600.0));
+            }
         });
     }
 
     TextInput {
+        id,
         data: ViewData::new(id),
-        cursor_glyph_idx: 0,
-        placeholder_text: None,
-        placeholder_buff: None,
-        placeholder_style: Default::default(),
-        selection_style: Default::default(),
-        buffer,
-        text_buf: None,
+        editor,
+        offset: 0,
```

Comment:
Not sure you need the `offset` field, can't you just use the editor's cursor offset?